internal/autoimpersonation/impersonation.go

```diff
@@ -0,0 +1,237 @@
+) (client.Writer, error) {
+	c := w.restConfig // shallow colpy of rest.Config
+	ic, err := w.impersonationConfigForObject(ctx, obj)
+	if err != nil {
+		return nil, err
+	}
+	c.Impersonate = ic
+	return client.New(&c, client.Options{})
+}
+
+func (w AutoImpersonatingWriterWrapper) impersonationConfigForObject(
+	ctx context.Context,
+	obj client.Object,
+) (rest.ImpersonationConfig, error) {
+	toImpersonate, err := w.getOwner(ctx, obj)
+	if err != nil {
+		return rest.ImpersonationConfig{}, fmt.Errorf("get owner for: %w", err)
+	}
+
+	user, groups := impersonationUserAndGroupsForObject(toImpersonate)
+	return rest.ImpersonationConfig{
+		UserName: user,
+		Groups:   groups,
+	}, nil
+}
+
+type objectIdentity struct {
+	Kind      string
+	Name      string
+	Namespace string
+}
+
+func (w AutoImpersonatingWriterWrapper) getOwner(ctx context.Context, obj client.Object) (objectIdentity, error) {
+	for _, ownerRef := range obj.GetOwnerReferences() {
+		gv, err := schema.ParseGroupVersion(ownerRef.APIVersion)
+		if err != nil {
+			return objectIdentity{}, err
+		}
+		if ownerRef.Controller != nil &&
+			*ownerRef.Controller &&
+			gv.Group == corev1alpha1.GroupVersion.Group {
+			// The controller is a Package Operator Object.
+			// -> check it's owners.
+			robj, err := w.scheme.New(schema.FromAPIVersionAndKind(ownerRef.APIVersion, ownerRef.Kind))
+			if err != nil {
+				return objectIdentity{}, err
+			}
+			potentialOwner := robj.(client.Object)
+
+			var ns string
+			if !strings.HasPrefix(ownerRef.Kind, "Cluster") {
+				ns = obj.GetNamespace() // same namespace as self.
+			}
+			if err := w.reader.Get(ctx, client.ObjectKey{
+				Name:      ownerRef.Name,
+				Namespace: ns,
+			}, potentialOwner); err != nil {
+				return objectIdentity{}, err
+			}
+			// recurse into this function
+			return w.getOwner(ctx, potentialOwner)
+		}
+	}
+	return objectIdentity{
+		Name:      obj.GetName(),
+		Namespace: obj.GetNamespace(),
+		Kind:      obj.GetObjectKind().GroupVersionKind().Kind,
+	}, nil
+}
+
+const impersonationPrefix = "pko"
+
+func impersonationUserAndGroupsForObject(oi objectIdentity) (user string, groups []string) {
+	var (
+		resourceSingular string
+		resourcePlural   string
+		isClusterScope   bool
+	)
+	switch oi.Kind {
+	case "ObjectSet":
+		resourceSingular = "objectset"
+		resourcePlural = "objectsets"
+	case "ObjectDeployment":
+		resourceSingular = "objectdeployment"
+		resourcePlural = "objectdeployments"
+	case "Package":
+		resourceSingular = "package"
+		resourcePlural = "packages"
+	case "ClusterObjectSet":
+		resourceSingular = "clusterobjectset"
+		resourcePlural = "clusterobjectsets"
+		isClusterScope = true
+	case "ClusterObjectDeployment":
+		resourceSingular = "clusterobjectdeployment"
+		resourcePlural = "clusterobjectdeployments"
+		isClusterScope = true
+	case "ClusterPackage":
+		resourceSingular = "clusterpackage"
+		resourcePlural = "clusterpackages"
+		isClusterScope = true
+	default:
```

Comment:
Hey @robshelly,
I took a quick peek and saw that there are `(Cluster)ObjectTemplate` cases missing here. 🙂